rfc/0000-batch-cancellation.rst

```diff
@@ -0,0 +1,538 @@
+Running -> {Success, Error, Failed, Cancelled}. The queue manager is
+responsible for updating the job state from Queued -> Ready within a
+loop. We bound the number of "Ready" jobs such that cancellation can
+happen on jobs one by one without requiring extensive bookkeeping on
+the number of ready jobs and their always_run state. Each user gets
+a background process that iterates through the "Queued" jobs that
+are not in cancelled batches or are set to always run. Each
+background process "submits" the update request to a new rate
+limiter that only allows 200 requests / second (maximum steady state
+scheduling rate) and limits the number of "Ready" jobs to a given
+number that can be cancelled within 30 seconds. The rate limiter
+uses the proportions of users with ready jobs to determine the
+frequency to allow their requests when rate limiting is in order. I
+have not decided on the exact algorithm for computing the fair share
+throttling proportions, but that should not be too hard to do. A
+crude example is a user with 0% of the cluster allocated to them
+should have 100 times more requests succeed than a user with 100% of
+the cluster in a simple two user example.
+
+With this model, we do not have to worry about tracking the number of
+cancellable ready jobs to be able to do an O(1) database update
+operation. Instead, the queue manager makes sure cancellation can
+happen within an acceptable amount of time by bounding the total
+number of "Ready" jobs. The benefits long-term of this model are we
+can have more complicated SQL queries for selecting the optimal job
+composition to satisfy constraints on regions, maximum number of
+concurrent jobs, and burn rate limits without having to have exact
+calculations for fair share (the rate limiter handles the fair
+share). The autoscaler SQL query can be very simple because it does
+not have to globally optimize the head of the job queue taking regions
+into account. The actual autoscaling algorithm can be more
+sophisticated as it just has to consider a subset of jobs that have
+been pre-selected. Finally, the scheduler can be significantly
+faster. The query that it uses just needs to find jobs that are
+"Ready" without worrying about fair share, cancellation, and always
+run. We can get better performance with higher throughputs on the
+scheduling query (we don't have to go batch by batch). We can also
+eventually use skip locks to have multiple scheduling loops going so
+we don't have to wait for 50 jobs to complete scheduling before
+querying the database for the next set of jobs that are ready to be
+scheduled.
+
+I think this option is better than the current proposed set of changes
+in this RFC. However, the cost of development for this approach is
+substantial and we need to do something simpler in the short term in
+order for QoB to be able to efficiently cancel a subset of jobs.
+
+
+-----------------------
+Extension to Job Groups
+-----------------------
```

Comment:
I don't think discussion of job groups is necessary for this RFC. If this is about a simplification of cancellation, such a change can stand on its own. I am personally more concerned about introducing complexity due to conflating separate ideas than I am about closing the door to future improvements.

Reply:
I think it's important to include this since its job groups that inspired this change. AFAIU, the motivation to change cancellation is additionally motivated by the complexity of cancelling a job group.